### PR TITLE
New resource - `azurerm_mssql_server_dns_alias`

### LINKED ISSUE #11276: Support for Azure SQL server DNS alias
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Add support for SQL Server DNS Aliases

### New or Affected Resource(s)

* azurerm_mssql_server_dns_alias

### Potential Terraform Configuration

```hcl
resource "azurerm_mssql_server_dns_alias" "alias" {
  name = "my_server"
  mssql_server_id = ...
  dns_record = "my_server.database.windows.net"
}
```

### References

* https://docs.microsoft.com/en-us/rest/api/sql/2021-08-01-preview/server-dns-aliases/create-or-update

### PR DESCRIPTION
**Pull request by wuxu92**
fix #11276

@magodo

### CODE REVIEW

File: internal/services/mssql/mssql_server_dns_alias_resource.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
+package mssql
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
```
Comment (magodo) [inline internal/services/mssql/mssql_server_dns_alias_resource.go:8]: There shouldn't import this package, all the exported entries are wrapped by `pluginsdk` instead.

File: internal/services/mssql/mssql_server_dns_alias_resource.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: validate.ServerID,
+		},
+
+		"name": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: webValidate.AppServiceEnvironmentName,
```
Comment (magodo) [inline internal/services/mssql/mssql_server_dns_alias_resource.go:40]: How can we use the validation function from app service to validate for this?

File: internal/services/mssql/mssql_server_dns_alias_resource.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+			if err := metadata.Decode(&alias); err != nil {
+				return err
+			}
+			serverID, err := parse.ServerID(alias.MsSQLServerId)
+			if err != nil {
+				return err
+			}
+			client := metadata.Client.MSSQL.ServerDNSAliasClient
+			subscriptionID := metadata.Client.Account.SubscriptionId
+			id := parse.NewServerDNSAliasID(subscriptionID, serverID.ResourceGroup, serverID.Name, alias.Name)
```
Comment (magodo) [inline internal/services/mssql/mssql_server_dns_alias_resource.go:76]: Can we do this instead?
```suggestion
			id := parse.NewServerDNSAliasID(serverID.SubscriptionId, serverID.ResourceGroup, serverID.Name, alias.Name)
```

File: internal/services/mssql/mssql_server_dns_alias_resource.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+			if err != nil {
+				return fmt.Errorf("creating %s: %v", id, err)
+			}
+			if err := future.WaitForCompletionRef(ctx, client.Client); err != nil {
+				return fmt.Errorf("waiting for creation of %s: %v", id, err)
+			}
+			metadata.SetID(id)
+			return nil
+		},
+		DiffFunc: nil,
```
Comment (magodo) [inline internal/services/mssql/mssql_server_dns_alias_resource.go:94]: This can be removed.
```suggestion
```

**Review by magodo — CHANGES_REQUESTED** [review]
Thanks for this PR!
I've taken a look through and left some comments inline, but this is mostly looking good to me 👍

File: internal/services/mssql/mssql_server_dns_alias_resource.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: validate.ServerID,
+		},
+
+		"name": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
+			ForceNew: true,
```
Comment (magodo) [inline internal/services/mssql/mssql_server_dns_alias_resource.go:37]: Can we set a `ValidateFunc` for this? At least ensure it is not an empty string.

**Review by magodo — CHANGES_REQUESTED** [review]
Thanks for this PR!
I have left some additional mostly minor comments that once addressed this should be good to merge 👍

File: internal/services/mssql/mssql_server_dns_alias_resource.go (hunk 5 above)
Comment (wuxu92) [reply internal/services/mssql/mssql_server_dns_alias_resource.go]: added a new validator for it, special characters and very long string are not acceptable by tests

File: internal/services/mssql/mssql_server_dns_alias_resource.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+			}
+			future, err := client.CreateOrUpdate(ctx, id.ResourceGroup, id.ServerName, id.DnsAliaseName)
+			if err != nil {
+				return fmt.Errorf("creating %s: %v", id, err)
+			}
+			if err := future.WaitForCompletionRef(ctx, client.Client); err != nil {
+				return fmt.Errorf("waiting for creation of %s: %v", id, err)
+			}
+			metadata.SetID(id)
+			return nil
```
Comment (katbyte) [inline internal/services/mssql/mssql_server_dns_alias_resource.go:89]: why is there no white space here? could we please make this more readable
```suggestion
			var alias ServerDNSAliasModel
			if err := metadata.Decode(&alias); err != nil {
				return err
			}
			
			serverID, err := parse.ServerID(alias.MsSQLServerId)
			if err != nil {
				return err
			}
			
			client := metadata.Client.MSSQL.ServerDNSAliasClient
			id := parse.NewServerDNSAliasID(serverID.SubscriptionId, serverID.ResourceGroup, serverID.Name, alias.Name)
			existing, err := client.Get(ctx, id.ResourceGroup, id.ServerName, id.DnsAliaseName)
			if !utils.ResponseWasNotFound(existing.Response) {
				if err != nil {
					return fmt.Errorf("retreiving %s: %v", id, err)
				}
				return metadata.ResourceRequiresImport(m.ResourceType(), id)
			}
			
			future, err := client.CreateOrUpdate(ctx, id.ResourceGroup, id.ServerName, id.DnsAliaseName)
			if err != nil {
				return fmt.Errorf("creating %s: %v", id, err)
			}
			if err := future.WaitForCompletionRef(ctx, client.Client); err != nil {
				return fmt.Errorf("waiting for creation of %s: %v", id, err)
			}
			
			metadata.SetID(id)
			return nil
```

File: internal/services/mssql/mssql_server_dns_alias_resource.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			var alias ServerDNSAliasModel
+			if err := metadata.Decode(&alias); err != nil {
+				return err
+			}
+			serverID, err := parse.ServerID(alias.MsSQLServerId)
+			if err != nil {
+				return err
+			}
+			client := metadata.Client.MSSQL.ServerDNSAliasClient
```
Comment (katbyte) [inline internal/services/mssql/mssql_server_dns_alias_resource.go:72]: this is always at the start of the function

**Review by katbyte — CHANGES_REQUESTED** [review]

**Review by katbyte — APPROVED** [review]
LGTM 🖲️

**Pull request merged**